### ISSUE TITLE
Pipeline reaper

### ISSUE THREAD

**Issue by evanchaoli** [concourse/concourse#4578]
## What challenge are you facing?

On some clusters, especially the experimental and staging clusters, we have frequently seen that some users deploy some pipelines and forget them, which leaves garbage pipelines on Concourse, making UI messy, wasting ATC's resource. 

## What would make this better?

I'm thinking to add a time-based pipeline reaper:

1. Add a CLI option `pipeline-reap-days`, default to 0. If set to 0, then disable the reaper.
2. When `pipeline-reap-days` is set to X days, if a pipeline has no new build triggered within X days, then reaper will pause the pipeline.
3. If a pipeline has been paused for X days, then reaper will destroy the pipeline.

## Are you interested in implementing this yourself?

Yes. But I'd like to get an indication of go from Concourse dev team before starting to work on the feature.

**Comment by wagdav** [comment]
This proposal is very similar to #3281

**Comment by evanchaoli** [comment]
Thank @wagdav for the info. Yes they are similar,  the differences are:

* I don't want to rely users to set TTL with `set-pipeline`, instead, if no build pipeline will be force paused and destroyed.
* With TTL, even if a pipeline is still generating new build, it will still be destroyed once TTL expires. But in my proposal, reaper will only destroy no-build pipelines.

But I think the reaper can also handle TTL.

**Comment by vito** [comment]
I like the idea! How would you feel about archiving the pipelines, rather than deleting them? That will still get them out of the way, hiding them in the UI and making the pipeline completely inactive on the backend, while also allowing a user to un-archive it if they want it back after all. (Maybe they were on vacation?)

https://github.com/concourse/rfcs/pull/33

We'd need to implement pipeline archiving first, but I don't think it's too much work compared to some of the other things on the roadmap to v10. 🙂 @tkellen started on it here before I even wrote the RFC: https://github.com/concourse/concourse/pull/2619 - although it has since stalled. If anyone's willing to kick-start it, we're happy to assist with feedback/advice!

Otherwise, we could start on this and just switch it to archiving once it's implemented later. Either way is fine. 🙂

**Comment by evanchaoli** [comment]
Archiving pipeline is a good idea, but based on our use cases, I don't feel it's super useful. Because as long as a pipeline yaml file is there, it's easy to set-pipeline line, archiving would only retain a build history, build logs should have been reaped already. Also archiving may add a new state to pipeline, in turn adds more complexities. Comparing gains with burdens, I at least archiving is not a high priority feature to implement.

In the mean time, pipeline reaper is not a high priority feature either. So I'm going to wait until I get bandwidth. At then, if archiving has been implemented, then we can archive no-build-for-long-time pipelines, otherwise do pending-destroy.

**Comment by vito** [comment]
(bumping since there's an associated PR in flight)

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.

**Comment by taylorsilva** [comment]
In v7.7.0 a pipeline pauser was added https://github.com/concourse/concourse/pull/7876

Not exactly the same, but does save on compute within a cluster.
